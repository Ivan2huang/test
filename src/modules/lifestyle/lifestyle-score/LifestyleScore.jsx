import React, { useEffect } from 'react';
import { Box, Card, Hidden, withStyles } from '@material-ui/core';
import { injectIntl } from 'react-intl';
import * as PropTypes from 'prop-types';
import Typography from '../../../uiComponents/Typography';

import { formatMessage } from '../../../helpers/helpers';
import GridItem from '../../../uiComponents/GridItem';
import Grid from '../../../uiComponents/Grid';
import LifestyleScoreChart from './LifestyleScoreChart';
import ComponentLoaderAndError from '../../common/shared/ComponentLoaderAndError';

const Styles = theme => ({
  scoreCard: {
    height: '100%',
    boxSizing: 'border-box',
    padding: `${theme.spacingX(6)}`,
    background: theme.white,
    boxShadow: '0px 2px 6px rgba(0, 0, 0, 0.05)',
    '&:last-child': {
      paddingBottom: theme.spacingX(6),
    },
    [theme.breakpoints.up('md')]: {
      padding: `${theme.spacingX(12)}`,
      '&:last-child': {
        paddingBottom: theme.spacingX(12),
      },
    },
  },
  scoreContainer: {
    textAlign: 'center',
    left: '50%',
    bottom: 'calc(50% - 16px)',
    transform: 'translate(-50%,50%)',
  },
});

const LifestyleScore = ({
  intl,
  classes,
  loading,
  errorState,
  healthScore,
  getLifestyleHealthScore,
}) => {
  const maxLifeStyleScore = 100;

  useEffect(() => {
    getLifestyleHealthScore();
  }, []);

  return (
    <Box component={Card} classes={{ root: classes.scoreCard }}>
      <Hidden implementation="css" smDown>
        <Typography type="style-7">
          {formatMessage(
            intl,
            'lifestyle.lifestyleScore.header',
            'My lifestyle score',
          )}
        </Typography>
      </Hidden>
      <Box mt={4}>
        <ComponentLoaderAndError loading={loading} errorState={errorState}>
          <Grid>
            <GridItem columns={{ md: 6 }}>
              <Box pt={{ xs: 2, md: 4 }}>
                <Typography type="style-6">
                  {formatMessage(
                    intl,
                    'lifestyle.lifestyleScore.description',
                    'Higher scores are associated with better lifestyle choices and lower risks of chronic disease.',
                  )}
                </Typography>
              </Box>
            </GridItem>
            <GridItem columns={{ xs: 12, md: 6 }}>
              <Box position="relative">
                <LifestyleScoreChart healthScore={healthScore} />
                <Box
                  position="absolute"
                  classes={{ root: classes.scoreContainer }}
                >
                  <Typography type="style-1" fontWeight="semiBold">
                    {formatMessage(
                      intl,
                      'lifestyle.lifestyleScore.healthScore',
                      `${healthScore}`,
                      {
                        healthScore,
                      },
                    )}
                  </Typography>
                  <Typography type="style-8">
                    {formatMessage(
                      intl,
                      'lifestyle.lifestyleScore.maximumScore',
                      `OUT OF ${maxLifeStyleScore}`,
                      {
                        maxScore: maxLifeStyleScore,
                      },
                    )}
                  </Typography>
                </Box>
              </Box>
            </GridItem>
            <GridItem columns={{ xs: 12 }}>
              <Box mb={{ xs: 2, md: 4 }} mt={2} textAlign="center">
                <Typography type="style-7">
                  {formatMessage(
                    intl,
                    'lifestyle.lifestyleScore.header',
                    'My lifestyle score',
                  )}
                </Typography>
              </Box>
              <Box textAlign="center">
                <Typography type="style-6">
                  {formatMessage(
                    intl,
                    'lifestyle.lifestyleScore.description',
                    'Higher scores are associated with better lifestyle choices and lower risks of chronic disease.',
                  )}
                </Typography>
              </Box>
            </GridItem>
          </Grid>
        </ComponentLoaderAndError>
      </Box>
    </Box>
  );
};

LifestyleScore.propTypes = {
  intl: PropTypes.shape({}).isRequired,
  classes: PropTypes.shape({
    scoreCard: PropTypes.string.isRequired,
    scoreContainer: PropTypes.string.isRequired,
  }).isRequired,
  loading: PropTypes.bool.isRequired,
  errorState: PropTypes.bool.isRequired,
  healthScore: PropTypes.number,
  getLifestyleHealthScore: PropTypes.func.isRequired,
};

LifestyleScore.defaultProps = {
  healthScore: null,
};

export default injectIntl(withStyles(Styles)(LifestyleScore));
