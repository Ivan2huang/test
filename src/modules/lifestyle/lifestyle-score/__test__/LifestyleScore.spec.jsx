/* eslint-disable no-undef,react/prop-types */
import React from 'react';
import { render } from '@testing-library/react';
import LifestyleScore from '../LifestyleScore';
import withTheme from '../../../../themes/withThemeProvider';
import withIntl from '../../../../i18n/withIntlProvider';

jest.mock('../../../../uiComponents/Grid', () => ({ children, ...rest }) => (
  <div>
    Grid Component
    <span>{mockPropsCapture(rest)}</span>
    {children}
  </div>
));

jest.mock(
  '../../../../uiComponents/GridItem',
  () => ({ children, ...rest }) => (
    <div>
      GridItem Component
      <span>{mockPropsCapture(rest)}</span>
      {children}
    </div>
  ),
);

jest.mock(
  '../../../../uiComponents/Typography',
  () => ({ children, ...rest }) => (
    <div>
      (Typography)
      <span>{mockPropsCapture(rest)}</span>
      {children}
    </div>
  ),
);

jest.mock('../../../../helpers/helpers', () => ({
  formatMessage: jest.fn((intl, id, defaultMessage) => defaultMessage),
}));

jest.mock(
  '../../../common/shared/ComponentLoaderAndError',
  () => ({ children, ...rest }) => (
    <div>
      ComponentLoaderAndError Component
      <span>{mockPropsCapture(rest)}</span>
      {children}
    </div>
  ),
);

jest.mock('../LifestyleScoreChart', () => props => (
  <div>
    LifestyleScoreChart component
    <span>{mockPropsCapture(props)}</span>
  </div>
));

describe('LifestyleScore Component', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  const props = {
    healthScore: 42,
    getLifestyleHealthScore: jest.fn(),
    loading: true,
    errorState: true,
  };
  const Component = withTheme(withIntl(LifestyleScore));

  const setUp = (componentProps = props) => {
    return render(<Component {...componentProps} />);
  };

  it('should match snapshot', () => {
    const { container } = setUp();

    expect(container).toMatchSnapshot();
  });

  it('should call getLifestyleScore exactly once on first time render', () => {
    const { rerender } = setUp();

    expect(props.getLifestyleHealthScore).toHaveBeenCalledTimes(1);
    rerender(<Component {...props} />);
    expect(props.getLifestyleHealthScore).toHaveBeenCalledTimes(1);
  });
});
