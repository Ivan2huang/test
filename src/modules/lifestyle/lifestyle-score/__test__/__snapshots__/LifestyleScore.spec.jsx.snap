// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LifestyleScore Component should match snapshot 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation1 MuiCard-root LifestyleScore-scoreCard-1 MuiBox-root MuiBox-root-4 LifestyleScore-scoreCard-1 MuiPaper-rounded"
  >
    <div
      class="PrivateHiddenCss-smDown-38"
    >
      <div>
        (Typography)
        <span>
          type = style-7
        </span>
        My lifestyle score
      </div>
    </div>
    <div
      class="MuiBox-root MuiBox-root-48"
    >
      <div>
        ComponentLoaderAndError Component
        <span>
          loading = true
          errorState = true
        </span>
        <div>
          Grid Component
          <span />
          <div>
            GridItem Component
            <span>
              columns = [object Object]
            </span>
            <div
              class="MuiBox-root MuiBox-root-49"
            >
              <div>
                (Typography)
                <span>
                  type = style-6
                </span>
                Higher scores are associated with better lifestyle choices and lower risks of chronic disease.
              </div>
            </div>
          </div>
          <div>
            GridItem Component
            <span>
              columns = [object Object]
            </span>
            <div
              class="MuiBox-root MuiBox-root-50"
            >
              <div>
                LifestyleScoreChart component
                <span>
                  healthScore = 42
                </span>
              </div>
              <div
                class="MuiBox-root MuiBox-root-51 LifestyleScore-scoreContainer-2"
                classes="[object Object]"
              >
                <div>
                  (Typography)
                  <span>
                    type = style-1
                    fontWeight = semiBold
                  </span>
                  42
                </div>
                <div>
                  (Typography)
                  <span>
                    type = style-8
                  </span>
                  OUT OF 100
                </div>
              </div>
            </div>
          </div>
          <div>
            GridItem Component
            <span>
              columns = [object Object]
            </span>
            <div
              class="MuiBox-root MuiBox-root-52"
            >
              <div>
                (Typography)
                <span>
                  type = style-7
                </span>
                My lifestyle score
              </div>
            </div>
            <div
              class="MuiBox-root MuiBox-root-53"
            >
              <div>
                (Typography)
                <span>
                  type = style-6
                </span>
                Higher scores are associated with better lifestyle choices and lower risks of chronic disease.
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
